src/gradualizer_lib.erl

```diff
@@ -83,58 +84,105 @@ reverse_graph(G) ->
     {integer, erl_anno:new(0), 0};
-pick_value(?type(pos_integer)) ->
+pick_value(?type(pos_integer), _Types) ->
     {integer, erl_anno:new(0), 0};
-pick_value(?type(neg_integer)) ->
+pick_value(?type(neg_integer), _Types) ->
     {integer, erl_anno:new(0), -1};
-pick_value(?type(float)) ->
+pick_value(?type(float), _Types) ->
     {float, erl_anno:new(0), -1};
-pick_value(?type(atom)) ->
+pick_value(?type(atom), _Types) ->
     {atom, erl_anno:new(0), a};
-pick_value({atom, _, A}) ->
+pick_value({atom, _, A}, _Types) ->
     {atom, erl_anno:new(0), A};
-pick_value({ann_type, _, [_, Ty]}) ->
-    pick_value(Ty);
-pick_value(?type(union, [Ty|_])) ->
-    pick_value(Ty);
-pick_value(?type(tuple, any)) ->
+pick_value({ann_type, _, [_, Ty]}, Types) ->
+    pick_value(Ty, Types);
+pick_value(?type(union, [Ty|_]), Types) ->
+    pick_value(Ty, Types);
+pick_value(?type(tuple, any), _Types) ->
     {tuple, erl_anno:new(0), []};
-pick_value(?type(tuple, Tys)) ->
-    {tuple, erl_anno:new(0), [pick_value(Ty) || Ty <- Tys]};
-pick_value(?type(record, [{atom, _, RecordName}])) ->
+pick_value(?type(tuple, Tys), Types) ->
+    {tuple, erl_anno:new(0), [pick_value(Ty, Types) || Ty <- Tys]};
+pick_value(?type(record, [{atom, _, RecordName}]), _Types) ->
     {record, erl_anno:new(0), RecordName, []};
-pick_value(?type(record, [{atom, _, RecordName} | Tys])) ->
+pick_value(?type(record, [{atom, _, RecordName} | Tys]), Types) ->
     MFields = [
-        {record_field, erl_anno:new(0), {atom, erl_anno:new(0), FieldName}, pick_value(Ty)}
+        {record_field, erl_anno:new(0), {atom, erl_anno:new(0), FieldName}, pick_value(Ty, Types)}
         || ?type(field_type, [{atom, _, FieldName}, Ty]) <- Tys
     ],
     {record, erl_anno:new(0), RecordName, MFields};
-pick_value(?type(list)) ->
+pick_value(?type(list), _Types) ->
     {nil, erl_anno:new(0)};
-pick_value(?type(list,_)) ->
+pick_value(?type(list,_), _Types) ->
     {nil, erl_anno:new(0)};
-pick_value(?type(nil)) ->
+pick_value(?type(nil), _Types) ->
     {nil, erl_anno:new(0)};
 %% The ?type(range) is a different case because the type range
 %% ..information is not encoded as an abstract_type()
 %% i.e. {type, Anno, range, [{integer, Anno2, Low}, {integer, Anno3, High}]}
-pick_value(?type(range, [{_TagLo, _, neg_inf}, Hi = {_TagHi, _, _Hi}])) ->
-    %% pick_value(Hi);
+pick_value(?type(range, [{_TagLo, _, neg_inf}, Hi = {_TagHi, _, _Hi}]), _Types) ->
+    %% pick_value(Hi, Types);
     Hi;
-pick_value(?type(range, [Lo = {_TagLo, _, _Lo}, {_TagHi, _, _Hi}])) ->
-    %% pick_value(Lo).
-    Lo.
+pick_value(?type(range, [Lo = {_TagLo, _, _Lo}, {_TagHi, _, _Hi}]), _Types) ->
+    %% pick_value(Lo, Types).
+    Lo;
+pick_value({user_type, Anno, Name, Args} = UserTy, Types) ->
+    case get_user_type_definition(UserTy, Types) of
+        {ok, Ty} ->
+            pick_value(Ty, Types);
+        opaque ->
+            throw({opaque, user_type, Anno, {Name, length(Args)}});
```

Comment:
Idea: return `{var, Anno, '_'}` here, for now. Or generate a unique OpaqueValueNNN variable name using unique_integer or something.